Lambda support in VM: Wolfgang's PR plus some simplifications (removing mask, clarifying ops)

brmataptos opened the pull request:
## Description

This PR started with a cherry-pick commit of Wolfgang's draft PR 15171 = commit #d85b919. 

On top of that, I 
- Remove the Mask feature since it complicates review,
- Extend and rename the opcodes to make them more general and clear
    - Any function value can have arguments bound to it in advance,
       not requiring a defined function handle except in `LdFunction[Generic]`.
- Tried to avoid unnecessary changes, but may have accidentally done so.
    - It took me a while to realize that we do indeed need a type parameter to
       `EarlyBind` and `Invoke` because of the way the verifier works today.
- add some semantics comment to `file_format.rs` which I'd worked out previously.

The current operations are:
    - `LdFunction`, `LdFunctionGeneric` = generate a closure from a defined function
    - `EarlyBind` = bind more arguments to any closure
    - `Invoke` = call a closure with final arguments
 
As noted by Wolfgang in his commit, we are still lacking:
- function value representation and interpreter
- function value serialization and deserialization
- some connections with front-end

## How Has This Been Tested?

Added tests for feature gating.  Should be safe to merge.

## Key Areas to Review

Did I lose any verifier support when modfiying Wolfgang's opcodes?

brmataptos commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/15387?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#15387** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/15387?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/15387?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#15365** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/15365?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#14792** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/14792?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#14791** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/14791?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#14742** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/14742?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

georgemitenkov commented on third_party/move/move-binary-format/src/file_format_common.rs:
```diff
@@ -789,6 +794,11 @@ pub fn instruction_key(instruction: &Bytecode) -> u8 {
+        // Since bytecode version 8
+        LdFunction(_) => Opcodes::LD_FUNCTION,
+        LdFunctionGeneric(_) => Opcodes::LD_FUNCTION_GENERIC,
+        Invoke(_) => Opcodes::INVOKE,
```
Probably we want InvokeFunction, EarlyBindFunction for consistency?

brmataptos replied:
Done.

Though I wonder if we should rename `Call` as `CallFunctionHandle` for similar consistency. :-)

brmataptos commented on third_party/move/move-vm/test-utils/src/gas_schedule.rs:
```diff
@@ -876,13 +883,23 @@ pub fn bytecode_instruction_costs() -> Vec<(Bytecode, GasCost)> {
 
 pub static INITIAL_COST_SCHEDULE: Lazy<CostTable> = Lazy::new(|| {
-    let mut instrs = bytecode_instruction_costs();
-    // Note that the DiemVM is expecting the table sorted by instruction order.
```
Note that `new_from_instructions` sorts the parameter the same way, so no need to sort again here.

georgemitenkov commented on third_party/move/move-binary-format/src/constant.rs:
```diff
@@ -17,6 +17,10 @@ fn sig_to_ty(sig: &SignatureToken) -> Option<MoveTypeLayout> {
         SignatureToken::U256 => Some(MoveTypeLayout::U256),
         SignatureToken::Vector(v) => Some(MoveTypeLayout::Vector(Box::new(sig_to_ty(v.as_ref())?))),
+        SignatureToken::Function(..) => {
+            // TODO(LAMBDA): do we need representation in MoveTypeLayout?
```
We do not need it here (function values cannot be constants), but we do need it otherwise. I have a PR adding it.

georgemitenkov replied:
Well, for structs it is also possible but we did not do it yet, so None is a way to go :)

georgemitenkov commented on third_party/move/move-binary-format/src/file_format.rs:
```diff
@@ -2934,6 +2969,134 @@ pub enum Bytecode {
+            assert ty == locals[#args -  i - 1]
+    "#]
+    #[gas_type_creation_tier_1 = "closure_ty"]
+    InvokeFunction(SignatureIndex),
```
Why? `LdFunction` does not need to load until use? We store idx/name in a function value, and only when we call `Invoke` we resolve that index/name and load the function?

brmataptos replied:
The `SignatureIndex` here is a static type, and is needed for verifier to allow the stack depth checker to know how many parameters/results are being popped/pushed on the stack, since the stack depth analysis is independent of the type analysis in the verifier (if these 2 analyses were combined, then this need not be included in the instruction, as it should be easy to derive from context; I really don't know why there are bunch of independent passes in the verifier).

Execution will have to look at the function value (1) to get the early-bound parameters to pass, and (2) to get the FunctionHandle/FunctionInstantiation to (a) get its type to figure out the number and type of parameters needed to pop off the stack for this call, and (b) determine which function is being called.

The dynamic types will have to be checked at some point, either when the value is created (LdFunction/EarlyBind/Deserialization) or called (InvokeFunction).  If we think it will be called at least once, then earlier type-checking would make sense, but if pulling the function data into cache is the slow part of type checking, we might want to do it at Call time.

brmataptos commented on third_party/move/move-binary-format/src/serializer.rs:
```diff
@@ -800,6 +800,9 @@ fn serialize_signature_token_single_node_impl(
             serialize_type_parameter_index(binary, *idx)?;
         },
+        SignatureToken::Function(..) => {
+            unimplemented!("serialization of function types")
```
Coming in a later PR.

brmataptos commented on third_party/move/move-binary-format/src/file_format.rs:
```diff
@@ -2934,6 +2969,134 @@ pub enum Bytecode {
+        func param types match provided parameter types
+    "#]
+    #[gas_type_creation_tier_0 = "closure_ty"]
+    EarlyBindFunction(SignatureIndex, u8),
```
Any function generated  for the body of an lambda expression can order free variables before lambda parameters so that it is natural to early-bind the `k` first parameters.  If the user wants to use an existing function with parameters in the wrong order, the user can add a wrapper function as necessary.

The mask idea and the code Wolfgang and I had written to deal with more arbitrary permutations of early/late-bound parameters was very "clever" and therefore hard to review and really an unnecessary security risk.

brmataptos replied:
replied already

brmataptos commented on third_party/move/move-bytecode-verifier/src/acquires_list_verifier.rs:
```diff
@@ -202,6 +210,29 @@ impl<'a> AcquiresVerifier<'a> {
+        // TODO(LAMBDA) In the future this may change.
+        let function_handle = self.module.function_handle_at(fh_idx);
+        let function_acquired_resources = self.function_acquired_resources(function_handle, fh_idx);
+        if !function_acquired_resources.is_empty() {
```
We disallow both reentrancy and nonempty acquires lists for now, so it's definitely safe, if a bit limited.  Your foreign move module can't call back into this one, since you're already on the stack.

In the future we can add more general access_specifiers to functions and function types, and check that the dynamic call doesn't exceed the accesses of the caller.  I'd rather work that out first, but people are anxious about an MVP deadline.

brmataptos replied:
I was being overly cautious here.   You're right, as long as we disallow all dynamic reentrancy we can ignore the acquires list of the function here.

Let's discuss post-MVP approaches to loosen constraints later.

brmataptos commented on third_party/move/move-binary-format/src/file_format.rs:
```diff
@@ -2935,6 +2970,134 @@ pub enum Bytecode {
+            assert ty == locals[#args -  i - 1]
+    "#]
+    #[gas_type_creation_tier_1 = "closure_ty"]
+    Invoke(SignatureIndex),
```
I don't think that is useful, as it is not really a pointer.  I'd be tempted to rename it `Apply` as a nod to Lisp's nomenclature, but I don't think it really matters here.  Does `Invoke` confuse you?

brmataptos commented on third_party/move/move-binary-format/src/file_format.rs:
```diff
@@ -2934,6 +2969,134 @@ pub enum Bytecode {
+        ..
+        stack >> arg_k
+
+        old_conservative_invocation_mode = conservative_invocation_mode
```
For LAMBDA MVP, we are playing it very safe and disallowing any call to module `m` if there is any dynamic call on the stack.  This is enforced by a VM state flag `conservative_invocation_mode`, which is set if we are in a dynamic call.

This is because dynamic calls violate the static dependency guarantees.  After we have the MVP implemented we will see how far we can prove dynamic safety, e.g., using access specifiers.

There was an error here in that the dynamic call also needs to always check that its module is not already on the stack.  Fixed.

brmataptos commented on third_party/move/move-bytecode-verifier/src/acquires_list_verifier.rs:
```diff
@@ -202,6 +210,29 @@ impl<'a> AcquiresVerifier<'a> {
+        Ok(())
+    }
+
+    fn invoke_acquire(&mut self, _offset: CodeOffset) -> PartialVMResult<()> {
```
Not sure what you mean by "nil-op" here.  Did you mean what I have done here?

brmataptos reviewed (commented):
Addressed comments, thanks.

georgemitenkov commented on third_party/move/move-bytecode-verifier/src/dependencies.rs:
```diff
@@ -455,6 +455,18 @@ fn compare_types(
+            compare_cross_module_signatures(context, args1, args2, def_module)?;
+            compare_cross_module_signatures(context, result1, result2, def_module)?;
+            if ab1 != ab2 {
+                Err(PartialVMError::new(StatusCode::TYPE_MISMATCH))
```
Well, we can use debug and format via "{:?}"? Any extra information is better than a status code. Imagine you have debug a type mismatch - there are so many places it can come from, it is difficult to bisect, error messages allows us to pinpoint the location (if they are propagated correctly 😄)

Below is some old code I do not know who wrote, but I would love to see error messages as well there.

georgemitenkov commented on third_party/move/move-binary-format/src/file_format.rs:
```diff
@@ -1456,6 +1482,13 @@ impl SignatureToken {
         }
     }
 
+    /// Returns true if the `SignatureToken` is a function.
```
I disagree: off-line discussions will lead to nothing because no one will do the scan across all aptos-core to fix wrong/outdated comments or their style. We should fix these things incrementally as we touch the code.

For this specific issue, I am fine with keeping `` for now. To give you an example:
```
/// `InterpreterImpl` instances can execute Move functions.
///
/// An `Interpreter` instance is a stand alone execution context for a function.
/// It mimics execution on a single thread, with an call stack and an operand stack.
pub(crate) struct InterpreterImpl { ... 
```

brmataptos replied:
Anyone who cares can do the scan and fix.  It's much easier to review a style cleanup as an independent PR.

georgemitenkov commented on third_party/move/move-binary-format/src/file_format.rs:
```diff
@@ -1389,19 +1410,22 @@ impl std::fmt::Debug for SignatureToken {
 
 impl SignatureToken {
-    /// Returns true if the token is an integer type.
+    // Returns `true` if the `SignatureToken` is an integer type.
```
I would prefer to have less changed lines in diff, but up to you

brmataptos commented:
@georgemitenkov I've added feature gating and tests to verify that it's being done.  I see some is done in e2e-tests but I can't do it there because the compiler doesn't have complete codegen yet.  I've added a test in `third_party/move/move-bytecode-verifier/bytecode-verifier-tests/src/unit_tests/feature_function-values_tests.rs` which manually constructs the bytecode file.  Fun!

PTAL.

brmataptos commented on third_party/move/move-binary-format/src/file_format.rs:
```diff
@@ -1254,6 +1261,8 @@ pub enum SignatureToken {
     /// Vector
     Vector(Box<SignatureToken>),
+    /// Function, with n argument types and m result types, and an associated ability set.
+    Function(Vec<SignatureToken>, Vec<SignatureToken>, AbilitySet),
```
Using a struct-style enum variant here causes those mysterious Clippy problems.  I think it's related to one of the `derive` macros on `SignatureToken` declaration (probably `dearbitrary::Dearbitrary`) lacking handling of struct-style enum variant.  In the interest of moving faster, I'm going back to the tuple-style enum variant instead, rather than trying to fix that 3d-party code. :-)

georgemitenkov commented on third_party/move/move-bytecode-verifier/src/stack_usage_verifier.rs:
```diff
@@ -228,6 +228,50 @@ impl<'a> StackUsageVerifier<'a> {
+
+            // EarlyBindFunction pops a function value and the captured arguments and returns 1 value
+            Bytecode::EarlyBindFunction(idx, arg_count) => {
+                if let Some(SignatureToken::Function(args, ..)) =
```
nit:
```
if matches!(
  self.resolver.signature_at(*idx).0.first(),
  Some(SignatureToken::Function(args, ..)) if args.len() <= *arg_count as usize
)
```

georgemitenkov commented on third_party/move/move-bytecode-verifier/src/type_safety.rs:
```diff
@@ -300,6 +301,152 @@ fn call(
+    count: u8,
+) -> PartialVMResult<()> {
+    let count = count as usize;
+    let SignatureToken::Function(param_tys, ret_tys, abilities) = expected_ty else {
```
For code readability, I would say it is better to match?
```
let closure_ty = safe_unwrap!(verifier.stack.pop());
match (closure_ty, expected_ty) {
  (Function(p, r, a), Function(pp, rr, aa)) => {
    // checks for inner args here
  },
  _ => // error
}
```

georgemitenkov commented on third_party/move/move-bytecode-verifier/src/signature_v2.rs:
```diff
@@ -881,6 +900,34 @@ impl<'a, const N: usize> SignatureChecker<'a, N> {
+                        checked_early_bind_insts.entry((*idx, *count))
+                    {
+                        // Note non-function case is checked in `verify_fun_sig_idx` above.
+                        if let Some(SignatureToken::Function(args, ..)) =
```
nit: you can use `if matches!(Some(...) if condition) { .. }`

georgemitenkov reviewed (commented):
Full pass. Do we have a way to test the verifier changes? I saw tests for feature verification, but I do not know where we test other passes.

brmataptos commented on third_party/move/move-binary-format/src/file_format.rs:
```diff
@@ -1010,6 +1010,13 @@ impl AbilitySet {
         self.0
     }
+
+    pub fn to_string_concise(self) -> String {
```
I didn't want to change all existing outputs to look different and make review difficult.  This is used for ability display just for functions.
